src/transformers/models/bloom/configuration_bloom.py

```diff
@@ -0,0 +1,167 @@
+            The dropout probability for all fully connected layers in the embeddings, encoder, and pooler.
+        embd_pdrop (`int`, *optional*, defaults to 0.1):
+            The dropout ratio for the embeddings.
+        attn_pdrop (`float`, *optional*, defaults to 0.1):
+            The dropout ratio for the attention.
+        layer_norm_epsilon (`float`, *optional*, defaults to 1e-5):
+            The epsilon to use in the layer normalization layers.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        apply_residual_connection_post_layernorm (`bool`, *optional*, defaults to `False`):
+            If enabled, use the layer norm of the hidden states as the residual in the transformer blocks
+        bias_dropout_fusion (`bool`, *optional*, defaults to `True`):
+            If enabled, apply dropout when adding the attention output together with the attention bias in the
+            transformer blocks
+        skip_bias_add (`bool`, *optional*, defaults to `True`):
+            If set to `True`, it will skip bias add for each linear layer in the transformer blocks
+        skip_bias_add_qkv (`bool`, *optional*, defaults to `False`):
+            If set to `True`, it will skip bias add for the first linear layer in the transformer blocks
+        attention_softmax_in_fp32 (`bool`, *optional*, defaults to `True`):
+            If set to `True` and the `dtype` is set to `float16` it will scale the input of the Softmax function to
+            `fp32`
+        hidden_dropout (`float`, *optional*, defaults to 0.1):
+            Dropout rate of the dropout function in `bias_dropout_fusion`
+        attention_dropout (`float`, *optional*, defaults to 0.1):
+            Dropout rate applied to the attention probs
+        scale_attn_weights (`bool`, *optional*, defaults to `True`):
+            Scale attention weights by dividing by sqrt(hidden_size)..
+        use_cache (`bool`, *optional*, defaults to `True`):
+            Whether or not the model should return the last key/values attentions (not used by all models).
+        dtype (`str`, *optional*, defaults to `"bfloat16"`):
+            Precision that has been used for the model's training in Megatron. Please load the model in the correct
+            precision by doing `model = BloomModel.from_pretrained(model_name, torch_dtype="auto")`.`
+        scale_attn_by_inverse_layer_idx (`bool`, *optional*, defaults to `False`):
+            Whether to additionally scale attention weights by `1 / layer_idx + 1`.
+        reorder_and_upcast_attn (`bool`, *optional*, defaults to `False`):
+            Whether to scale keys (K) prior to computing attention (dot-product) and upcast attention
+            dot-product/softmax to float() when training with mixed precision.
+        pretraining_tp (`int`, *optional*, defaults to `1`):
```

Comment:
Yes you are right, all TP artifacts are indeed dropped when converting the model. 
But I have noticed in my experiments that to obtain exact match of RowParallel layers logits, you need to manually reproduce some slicing operations. Please see: https://github.com/pytorch/pytorch/issues/76232 
We can discard that but we lose a bit in terms of exactness

Reply:
As I commented here https://github.com/huggingface/transformers/pull/17202#discussion_r883109242
this leads to very inefficient runtime.

I wonder if it'd be much more efficient to do a short finetuning with TP=1 when the 176B model is done. 

This needs to be done once and everybody will benefit from it, but if you code for exactness the slowdown will impact thousands of users multiplied by thousands of hours.

Reply:
I see the point! Maybe the best tradeoff for now (since we are also playing with intermediate checkpoints) is to have a boolean in the config file (for eg exact_op) that activates the manual tp aggregation. Users can use it at "their own risk" until we fine tune with TP=1 as you suggest.
Then the idea is to progressively remove this extra switch once we fine-tune with TP=1 at the end of the training.

Reply:
Why do you need exactness for for intermediate checkpoints? To compare lm-eval results?

As I was trying to hypothesize above it'd be great to actually figure out the tolerance and when the model is underperforming because of what you're trying to overcome.

Also don't forget that the whole TP many->one conversion is very approximate since we average the rows - so it's not precise anyway.

Reply:
Perhaps we could run eval-harness on the exact-but-slow vs. close-but-fast version and check qualitative results (softmax) rather than logits' exactness?

Reply:
Agreed with that!
Though I have few questions:
- You mean running eval harness on 176b?
- Do you think it can be done after merging ? (ie by using the switch)

Reply:
@younesbelkada, I think this is not the best place to discuss this potentially crucial choice - could you please bring it up in the modeling group on slack? as it has the right people to make the best decision.

Reply:
>    You mean running eval harness on 176b?

Yes

>    Do you think it can be done after merging ? (ie by using the switch)

I'm not sure. 

I realize this PR has been going on for many weeks and it'd be great to merge it. 

Perhaps marking the slow version as experimental and that it could be removed at any point? Because once you merge you have to maintain backward compatibility...

Perhaps let's have a limited time span discussion on slack and say by Monday make a decision and go with that?

Reply:
yeah let me post a message on the modeling group!

Reply:
Ok seems that we are converging towards having a classic linear layer if I understood well from the thread on Slack. By default we can set the TP to 1 and we should keep it so that we will be able to compare it on eval harness for qualitative results

Reply:
when performance benchmarking is done ideally you want to have enough load for at least a few secs in the final report. Otherwise often the overhead of measurement makes the differences very difficult to see. And you want to measure more than one sequence as a single or a few results don't quite give you the right picture. So multiple rounds and then averaging the results.

> TP=4 does not seem to be that slower than TP=1

If you're measuring the ensemble, of course overall the difference will be small. If you were to measure just the parts that we are discussing and the inputs are realistic and many you should see the difference. Those things just add up when you do more than a single inference on a short input, that's why we bother with making small sections of code faster.

> What scares me is the greatest relative difference

I typically have a hard time making sense of that relative side of results in `assert_close`. May be someone else could answer that one.

Reply:
Ok so I have benchmarked the specific portion we are discussion and it appears that in average the TP=4 is ~3.7 slower than TP=1 (which makes sense actually since we do 4 for loops).
For now I am confident to let TP=1 by default on all Bloom models (as it reproduced the same generations as TP=4 for now) - so qualitatively we are fine - but quantitatively it may affect some proper evaluation metrics. Therefore what I propose is to make the slow feature experimental as you suggested before.

> Perhaps marking the slow version as experimental and that it could be removed at any point? Because once you merge you have to maintain backward compatibility...

And I will put an explicit explanation of how to properly handle this argument in the docs. We will put TP=1 by default to let users benefit for the most efficient inference time. Also I believe that this feature will disappear in the future. I feel from the slack discussion that we are converging toward a successful finetuning with TP=1 once the training has been done.
I will push a commit to apply these modifications!

Reply:
I will resolve the conversation here, let's jump in the other PR to discuss further in case we need to!